addons/payment/controllers/portal.py

```diff
@@ -1,359 +1,418 @@
+            PaymentPostProcessing.remove_transactions(tx_sudo)
+            render_values = {
+                'tx': tx_sudo,
+                'status': status,
+                'message': message
+            }
+            return request.render('payment.confirm', render_values)
+        else:
+            # Display the portal homepage to the user
+            return request.redirect('/my/home')
 
-        reference_values = order_id and {'sale_order_ids': [(4, order_id)]} or {}
-        reference_values.update(acquirer_id=int(acquirer_id))
-        values['reference'] = request.env['payment.transaction']._compute_reference(values=reference_values, prefix=reference)
-        tx = request.env['payment.transaction'].sudo().with_context(lang=None).create(values)
-        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
-        token_str = '%s%s%s' % (tx.id, tx.reference, float_repr(tx.amount, precision_digits=tx.currency_id.decimal_places))
-        token = hmac.new(secret.encode('utf-8'), token_str.encode('utf-8'), hashlib.sha256).hexdigest()
-        tx.return_url = '/website_payment/confirm?tx_id=%d&access_token=%s' % (tx.id, token)
+    @http.route('/payment/validation', type='http', methods=['GET'], auth='user', website=True)
+    def payment_validation_transaction(self, tx_id, access_token, **kwargs):
+        """ Refund a validation transaction and redirect the user to the landing route.
 
-        PaymentProcessing.add_payment_transaction(tx)
+        :param str tx_id: The validation transaction, as a `payment.transaction` id
+        :param str access_token: The access token used to verify the user
+        :param dict kwargs: Locally unused data passed to `_refund_validation_transaction`
+        """
+        # Raise an HTTP 404 if the tx id or the access token is invalid
+        tx_id, = self.cast_as_numeric([tx_id], numeric_type='int')
+        tx = request.env['payment.transaction'].browse(tx_id).exists()
+        db_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        if not tx or not payment_utils.check_access_token(
+            access_token,
+            db_secret,
+            tx.partner_id.id,
+            tx.amount,
+            tx.currency_id.id
+        ):
+            raise werkzeug.exceptions.NotFound  # Don't leak info about existence of an id
+
+        landing_route = self._refund_validation_transaction(tx_id=tx_id, **kwargs)
+        return request.redirect(landing_route)
+
+    def _refund_validation_transaction(self, tx_id, **kwargs):
+        """ Refund a validation transaction and remove it from post-processing.
+
+        :param str tx_id: The validation transaction to refund, as a `payment.transaction` id
+        :param dict kwargs: Optional data. This parameter is not used here
+        :return: The landing route of the transaction
+        :rtype: str
+        :raise: ValidationError if the transaction id is invalid
+        """
```

Comment:
I made this method return the tx instead. I could make it return None but I'd prefer to keep this little 'hack' to avoid clone the code of the method in every validation route method. This way, the refund method handles the tx completely and makes all the necessary checks on it.